macros/src/bitflags.rs

```diff
@@ -0,0 +1,153 @@
+    let ts2 = proc_macro2::TokenStream::from(ts.clone());
+    let input: Input = parse_macro_input!(ts);
+
+    let codegen = Codegen::new(&input);
+
+    // Encode package and disambiguator to provide the decoder with all info it needs (even if
+    // technically redundant, since it's also stored in the symbol we create).
+    let format_str = format!(
+        "{{={}:__internal_bitflags_{}@{}@{}}}",
+        input.ty.to_token_stream(),
+        input.ident,
+        crate::symbol::package(),
+        crate::symbol::disambiguator(),
+    );
+    let sym = mksym(&format_str, "bitflags", false);
+
+    let ident = &input.ident;
+    let ty = input.ty;
+    let flags = codegen.flags.iter().map(|f| &f.def);
+    let result = quote! {
+        const _: () = {
+            fn assert<T: defmt::export::UnsignedInt>() {}
+            assert::<#ty>;
+
+            #(#flags)*
+        };
+
+        defmt::export::bitflags! {
+            #ts2
+        }
+
+        impl defmt::Format for #ident {
+            fn format(&self, f: defmt::Formatter) {
+                unreachable!()
+            }
+            fn _format_tag() -> defmt::Str {
+                #sym
+            }
+            fn _format_data(&self) {
+                // There's a method available for every supported bitflags type.
+                defmt::export::#ty(&self.bits());
+            }
+        }
+    };
+    result.into()
+}
+
+struct FlagStatic {
+    def: TokenStream2,
+}
+
+struct Codegen {
+    flags: Vec<FlagStatic>,
+}
+
+impl Codegen {
+    fn new(input: &Input) -> Self {
+        let flags = input
+            .flags
+            .iter()
+            .map(|flag| {
+                let name = &flag.ident;
+                let value = &flag.value;
+                let repr_ty = &input.ty;
+
+                let sym = Symbol::new(
+                    "bitflags_value",
+                    &format!("{}::{}", input.ident, flag.ident),
+                )
+                .mangle();
+
+                let def = quote! {
+                    #[cfg_attr(target_os = "macos", link_section = ".defmt,end")]
+                    #[cfg_attr(not(target_os = "macos"), link_section = ".defmt.end")]
+                    #[export_name = #sym]
+                    static #name: u128 = {
+                        // NB: It might be tempting to just do `#value as u128` here, but that
+                        // causes a value such as `1 << 127` to be evaluated as an `i32`, which
+                        // overflows. So we instead coerce (but don't cast) it to the bitflags' raw
+                        // type, and then cast that to u128.
+                        let coerced_value: #repr_ty = #value;
+                        coerced_value as u128
+                    };
+                };
+                FlagStatic { def }
+            })
+            .collect::<Vec<_>>();
+
+        Codegen { flags }
+    }
+}
+
+#[allow(dead_code)]
+struct Input {
```

Comment:
it seems quite unfortunate to have to parse the `bitflags!` syntax but I don't see any other way: we need the flag names and values and there's no trait in bitflags that encodes that information (it wouldn't be possible to encode flag names in a trait).

Reply:
Yeah, this seems unavoidable. But I don't think it's a big issue, `syn` makes writing a parser for this extremely easy, and the `bitflags` syntax hasn't really changed much (if at all) since 1.0 was released.

Reply:
I think it'd be more worried about new syntax being added (w/o it being a breaking change) and then us needing to update our bitflags parser to accommodate that, than having to support a new, totally different bitflags 2.0 syntax.

it seems very unlikely that syntax will grow however given that the last (patch) release was 2 years ago so now I'm not too worried about this.